src/core/utils/components/TextTruncation.tsx

```diff
@@ -0,0 +1,71 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import { CommonProps } from '../props';
+
+const ELLIPSIS_CHAR = '…';
+
+export type TextTruncationProps = {
+  /**
+   * Text to truncate.
+   */
+  text: string;
+  /**
+   * Number of characters to leave at the end.
+   * @default 6
+   */
+  endCharsCount?: number;
+} & CommonProps;
+
+/**
+ * Truncates text with the ellipsis in the middle,
+ * leaving defined number of chars at the end.
+ */
+export const TextTruncation = (props: TextTruncationProps) => {
+  const { text, endCharsCount = 6, ...rest } = props;
+
+  const childRef = React.useRef<HTMLDivElement>(null);
+
+  const [truncatedText, setTruncatedText] = React.useState(text);
+  React.useLayoutEffect(() => {
+    setTruncatedText(text);
+  }, [text]);
+
+  React.useLayoutEffect(() => {
+    if (!childRef.current) {
+      return;
+    }
+
+    const parentElement = childRef.current.parentElement ?? childRef.current;
+    if (parentElement.scrollWidth <= parentElement.clientWidth) {
+      return;
+    }
+
+    const { paddingLeft, paddingRight } = getComputedStyle(parentElement);
+    const actualWidth =
+      parentElement.clientWidth -
+      parseFloat(paddingLeft) -
+      parseFloat(paddingRight);
+    const charWidth = Math.ceil(parentElement.scrollWidth / text.length);
+    const totalVisibleCount = Math.floor(actualWidth / charWidth);
```

Comment:
It takes average char width that might cause some random errors when there are wide chars and smaller ones that might cause double ellipsis - in the middle and in the end.